File: components/framework/src/kgr/Pipe.cpp
/*
  SPDX-License-Identifier: LGPL-3.0

  Copyright (c) 2019-2020 Jevgenijs Protopopovs

  This file is part of Sloked project.

  Sloked is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License version 3 as
  published by the Free Software Foundation.


  Sloked is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with Sloked.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "sloked/kgr/Pipe.h"

#include "sloked/core/Error.h"

namespace sloked {

    KgrAsyncPipe::KgrAsyncPipe(std::unique_ptr<KgrPipe> pipe)
        : pipe(std::move(pipe)), baseCallback{nullptr} {
        if (this->pipe) {
            this->pipe->SetMessageListener([this] { this->Notify(); });
        }
    }

    KgrAsyncPipe::~KgrAsyncPipe() {
        if (this->pipe) {
            this->pipe->SetMessageListener(nullptr);
        }
    }

    std::unique_ptr<KgrPipe> KgrAsyncPipe::ChangePipe(
        std::unique_ptr<KgrPipe> pipe) {
        std::unique_lock lock(this->mtx);
        auto currentPipe = std::move(this->pipe);
        if (currentPipe) {
            currentPipe->SetMessageListener(nullptr);
        }
        this->pipe = std::move(pipe);
        this->baseCallback = nullptr;
        this->callbacks = {};
        if (this->pipe) {
            this->pipe->SetMessageListener([this] { this->Notify(); });
        }
        return currentPipe;
    }

    bool KgrAsyncPipe::Valid() const {
        std::unique_lock lock(this->mtx);
        return this->pipe != nullptr;
    }

    KgrPipe::Status KgrAsyncPipe::GetStatus() const {
        std::unique_lock lock(this->mtx);
        if (this->pipe) {
            return this->pipe->GetStatus();
        } else {
            throw SlokedError("AsyncPipe: Invalid pipe");
        }
    }

    bool KgrAsyncPipe::Empty() const {
        std::unique_lock lock(this->mtx);
        if (this->pipe) {
            return this->pipe->Empty();
        } else {
            throw SlokedError("AsyncPipe: Invalid pipe");
        }
    }

    std::size_t KgrAsyncPipe::Count() const {
        std::unique_lock lock(this->mtx);
        if (this->pipe) {
            return this->pipe->Count();
        } else {
            throw SlokedError("AsyncPipe: Invalid pipe");
        }
    }

    void KgrAsyncPipe::Close() {
        std::unique_lock lock(this->mtx);
        if (this->pipe) {
            this->pipe->SetMessageListener(nullptr);
            this->pipe->Close();
            this->pipe = nullptr;
        } else {
            throw SlokedError("AsyncPipe: Invalid pipe");
        }
    }

    TaskResult<KgrValue> KgrAsyncPipe::Read() {
        TaskResultSupplier<KgrValue> supplier;
        std::unique_lock lock(this->mtx);
        if (this->pipe == nullptr) {
            throw SlokedError("AsyncPipe: Invalid pipe");
        }
        auto callback = [this, supplier](std::unique_lock<std::mutex> lock) {
            if (!this->pipe->Empty()) {
                this->callbacks.pop();
                lock.unlock();
                supplier.SetResult(this->pipe->Read());
            }
        };
        if (this->callbacks.empty()) {
            this->callbacks.push(callback);
            callback(std::move(lock));
        } else {
            this->callbacks.push(callback);
            this->callbacks.front()(std::move(lock));
        }
        return supplier.Result();
    }

    TaskResult<void> KgrAsyncPipe::Wait(std::size_t count) {
        TaskResultSupplier<void> supplier;
        std::unique_lock lock(this->mtx);
        if (this->pipe == nullptr) {
            throw SlokedError("AsyncPipe: Invalid pipe");
        }
        auto callback = [this, supplier,
                         count](std::unique_lock<std::mutex> lock) {
            if (this->pipe->Count() >= count) {
                this->callbacks.pop();
                lock.unlock();
                supplier.SetResult();
            }
        };
        if (this->callbacks.empty()) {
            this->callbacks.push(callback);
            callback(std::move(lock));
        } else {
            this->callbacks.push(callback);
            this->callbacks.front()(std::move(lock));
        }
        return supplier.Result();
    }

    void KgrAsyncPipe::SetMessageListener(std::function<void()> callback) {
        std::unique_lock lock(this->mtx);
        if (this->pipe) {
            this->baseCallback = std::move(callback);
        } else {
            throw SlokedError("AsyncPipe: Invalid pipe");
        }
    }

    void KgrAsyncPipe::Drop(std::size_t count) {
        std::unique_lock lock(this->mtx);
        if (this->pipe) {
            this->pipe->Drop(count);
        } else {
            throw SlokedError("AsyncPipe: Invalid pipe");
        }
    }

    void KgrAsyncPipe::DropAll() {
        std::unique_lock lock(this->mtx);
        if (this->pipe) {
            this->pipe->DropAll();
        } else {
            throw SlokedError("AsyncPipe: Invalid pipe");
        }
    }

    void KgrAsyncPipe::Write(KgrValue &&value) {
        std::unique_lock lock(this->mtx);
        if (this->pipe) {
            this->pipe->Write(std::forward<KgrValue>(value));
        } else {
            throw SlokedError("AsyncPipe: Invalid pipe");
        }
    }

    bool KgrAsyncPipe::SafeWrite(KgrValue &&value) {
        std::unique_lock lock(this->mtx);
        if (this->pipe) {
            return this->pipe->SafeWrite(std::forward<KgrValue>(value));
        } else {
            throw SlokedError("AsyncPipe: Invalid pipe");
        }
    }

    void KgrAsyncPipe::Notify() {
        std::unique_lock lock(this->mtx);
        if (this->callbacks.empty()) {
            auto callback = this->baseCallback;
            lock.unlock();
            if (callback) {
                callback();
            }
        } else {
            auto callback = this->callbacks.front();
            callback(std::move(lock));
        }
    }
}  // namespace sloked

File: src/components/script/lua/scripts/slokedlib/pipe.lua
local promise = require 'slokedlib/promise'

local Pipe = {}
Pipe.__index = Pipe

local PromisifiedPipe = {}
PromisifiedPipe.__index = PromisifiedPipe

function Pipe:new(pipe)
    local obj = {
        pipe=pipe,
        listeners={}
    }
    setmetatable(obj, self)
    obj.pipe:listen(function()
        if #obj.listeners > 0 then
            local callback = table.remove(obj.listeners, 1)
            callback()
        end
    end)
    return obj
end

function Pipe:promisify(pipe)
    return PromisifiedPipe:new(pipe)
end

function applyCallback(callback, that, fn, ...)
    local status, result = pcall(fn, that, ...)
    if status then
        callback(nil, result)
    else
        callback(result, nil)
    end
end

function Pipe:isOpen()
    return self.pipe:isOpen()
end

function Pipe:count()
    return self.pipe:count()
end

function Pipe:empty()
    return self.pipe:empty()
end

function Pipe:close(callback)
    sloked.sched:defer(function ()
        applyCallback(callback, self.pipe, self.pipe.close)
    end)
end

function Pipe:read(callback)
    if not self:empty() then
        applyCallback(callback, self.pipe, self.pipe.read)
    else
        self.listeners[#self.listeners + 1] = function()
            applyCallback(callback, self.pipe, self.pipe.read)
        end
    end
end

function Pipe:wait(callback, count)
    if self:count() >= count then
        applyCallback(callback, self.pipe, self.pipe.wait, count)
    else
        self.listeners[#self.listeners + 1] = function()
            self:wait(callback, count)
        end
    end
end

function Pipe:drop(callback, count)
    if self:count() >= count then
        applyCallback(callback, self.pipe, self.pipe.drop, count)
    else
        self.listeners[#self.listeners + 1] = function()
            self:drop(callback, count)
        end
    end
end

function Pipe:dropAll(callback)
    applyCallback(callback, self.pipe, self.pipe.dropAll)
end

function Pipe:write(callback, value)
    applyCallback(callback, self.pipe, self.pipe.write, value)
end

function PromisifiedPipe:new(pipe)
    local obj = {
        pipe=Pipe:new(pipe)
    }
    setmetatable(obj, self)
    return obj
end

function PromisifiedPipe:isOpen()
    return self.pipe:isOpen()
end

function PromisifiedPipe:count()
    return self.pipe:count()
end

function PromisifiedPipe:empty()
    return self.pipe:empty()
end

function PromisifiedPipe:close()
    return promise:promisify(self.pipe, self.pipe.close)
end

function PromisifiedPipe:read()
    return promise:promisify(self.pipe, self.pipe.read)
end

function PromisifiedPipe:wait(count)
    return promise:promisify(self.pipe, self.pipe.wait, count)
end

function PromisifiedPipe:drop(count)
    return promise:promisify(self.pipe, self.pipe.drop, count)
end

function PromisifiedPipe:dropAll()
    return promise:promisify(self.pipe, self.pipe.dropAll)
end

function PromisifiedPipe:write(value)
    return promise:promisify(self.pipe, self.pipe.write, value)
end

return Pipe